// @gutenye/ocr-common/build/models/ModelBase.js
import { Tensor } from 'onnxruntime-common';
export class ModelBase {
    options;
    #model;
    constructor({ model, options }) {
        this.#model = model;
        this.options = options;
    }
    async runModel({ modelData, onnxOptions = {}, }) {
        const input = this.#prepareInput(modelData);
        const outputs = await this.#model.run({
            [this.#model.inputNames[0]]: input,
        }, onnxOptions);
        const output = outputs[this.#model.outputNames[0]];
        return output;
    }
    #prepareInput(modelData) {
        const input = Float32Array.from(modelData.data);
        return new Tensor('float32', input, [1, 3, modelData.height, modelData.width]);
    }
    imageToInput(image, { mean = [0, 0, 0], std = [1, 1, 1] }) {
        const R = [];
        const G = [];
        const B = [];
        for (let i = 0; i < image.data.length; i += 4) {
            R.push((image.data[i] / 255 - mean[0]) / std[0]);
            G.push((image.data[i + 1] / 255 - mean[1]) / std[1]);
            B.push((image.data[i + 2] / 255 - mean[2]) / std[2]);
        }
        const newData = [...B, ...G, ...R];
        return {
            data: newData,
            width: image.width,
            height: image.height,
        };
    }
    debugImage(image, path) {
        const { debugOutputDir, isDebug } = this.options;
        if (!isDebug || !debugOutputDir) {
            return;
        }
        image.write(`${debugOutputDir}/${path}`);
    }
    async debugBoxImage(sourceImage, lineImages, path) {
        const { debugOutputDir, isDebug } = this.options;
        if (!isDebug || !debugOutputDir) {
            return;
        }
        const boxImage = await sourceImage.drawBox(lineImages);
        boxImage.write(`${debugOutputDir}/${path}`);
    }
}
